feat: date picker core

bart-krakowski opened the pull request:
## DatePicker Core

### Summary
This RFC proposes the creation of a DatePickerCore class that extends the CalendarCore class to include additional functionalities for managing date selection. This includes single-date, multiple-date, and range selection modes. The DatePickerCore is designed for use in various date picker applications where precise date management and selection are required.

### Motivation
The motivation behind this RFC is to provide a flexible and efficient core for date picker applications, enabling consistent and simplified management of date selection functionalities. By extending CalendarCore, this class leverages existing calendar functionalities while adding robust support for date selection, ensuring it can handle a variety of use cases such as single-date, multiple-date, and range selection.

```ts
export class DatePickerCore extends CalendarCore {
  constructor(options: DatePickerCoreOptions);
}
```

The DatePickerCore class extends CalendarCore to provide additional functionalities for managing date selection, including single-date, multiple-date, and range selection modes. This class is designed to be used in various date picker applications where precise date management and selection are required.

### Parameters

- `weekStartsOn?: number`
  - An optional number that specifies the day of the week that the calendar should start on. It defaults to 1 (Monday).
- `events?: TEvent[]`
  - An optional array of events that the calendar should display.
- `viewMode: ViewMode`
  - An object that specifies the initial view mode of the calendar.
- `locale?: Parameters<Temporal.PlainDate['toLocaleString']>['0']`
  - An optional string that specifies the locale to use for formatting dates and times.
- `minDate?: Temporal.PlainDate`
  - An optional date that specifies the minimum selectable date.
- `maxDate?: Temporal.PlainDate`
  - An optional date that specifies the maximum selectable date.
- `selectedDates?: Temporal.PlainDate[]`
  - An optional array of dates that are initially selected.
- `multiple?: boolean`
  - An optional boolean that allows multiple dates to be selected if set to true.
- `range?: boolean`
  - An optional boolean that allows a range of dates to be selected if set to true.

### Returns

- `getSelectedDates(): Temporal.PlainDate[]`
  - Returns an array of selected dates.
- `selectDate(date: Temporal.PlainDate): void`
  - Selects a date. Depending on the configuration, this can handle single, multiple, or range selections.
- `getDaysWithEvents(): Array<Day<TEvent>>`
  - Returns an array of days in the current period with their associated events.
- `getDaysNames(): string[]`
  - Returns an array of strings representing the names of the days of the week based on the locale and week start day.
- `changeViewMode(newViewMode: ViewMode): void`
  - Changes the view mode of the calendar.
- `goToPreviousPeriod(): void`
  - Navigates to the previous period based on the current view mode.
- `goToNextPeriod(): void`
  - Navigates to the next period based on the current view mode.
- `goToCurrentPeriod(): void`
  - Navigates to the current period.
- `goToSpecificPeriod(date: Temporal.PlainDate): void`
  - Navigates to a specific period based on the provided date.
- `updateCurrentTime(): void`
  - Updates the current time.
- `getEventProps(id: Event['id']): { style: CSSProperties } | null`
  - Retrieves the style properties for a specific event based on its ID.
- `getCurrentTimeMarkerProps(): { style: CSSProperties; currentTime: string | undefined }`
  - Retrieves the style properties and current time for the current time marker.
- `groupDaysBy(props: Omit<GroupDaysByProps<TEvent>, 'weekStartsOn'>): (Day<TEvent> | null)[][]`
  - Groups the days in the current period by a specified unit. The fillMissingDays parameter can be used to fill in missing days with previous or next month's days.


### Example Usage
```ts
import { DatePickerCore, Event } from '@tanstack/time';
import { Temporal } from '@js-temporal/polyfill';

const datePickerCore = new DatePickerCore({
  weekStartsOn: 1,
  viewMode: { value: 1, unit: 'month' },
  events,
  locale: 'en-US',
  minDate: Temporal.PlainDate.from('2024-01-01'),
  maxDate: Temporal.PlainDate.from('2024-12-31'),
  selectedDates: [Temporal.PlainDate.from('2024-06-10')],
  multiple: true,
});

// Get selected dates
const selectedDates = datePickerCore.getSelectedDates();
console.log(selectedDates);

// Select a date
datePickerCore.selectDate(Temporal.PlainDate.from('2024-06-15'));

// Change view mode to week
datePickerCore.changeViewMode({ value: 1, unit: 'week' });

// Navigate to the next period
datePickerCore.goToNextPeriod();

// Update current time
datePickerCore.updateCurrentTime();

// Get event properties
const eventProps = datePickerCore.getEventProps('1');
console.log(eventProps);

// Get current time marker properties
const currentTimeMarkerProps = datePickerCore.getCurrentTimeMarkerProps();
console.log(currentTimeMarkerProps);

// Group days by week
const groupedDays = datePickerCore.groupDaysBy({ days: daysWithSelection, unit: 'week' });
console.log(groupedDays);
```

bart-krakowski committed:
refactor: update isValidDate return type for improved type safety

bart-krakowski committed:
refactor: update isValidDate return type for improved type safety

bart-krakowski committed:
refactor: update isValidDate return type for improved type safety

bart-krakowski committed:
test: refactor tests for the parse util

bart-krakowski committed:
refactor: types for utils

bart-krakowski committed:
revert: isValidDate util

bart-krakowski committed:
Merge branch 'v0' into chore/vitest-config

bart-krakowski committed:
docs: architecture md file

bart-krakowski committed:
feat: useCalendar hook

bart-krakowski committed:
test: useCalendar tests

bart-krakowski committed:
docs: useCalendar hook

bart-krakowski committed:
docs: useCalendar hook

bart-krakowski committed:
test: useCalendar tests

bart-krakowski committed:
test: useCalendar tests

bart-krakowski committed:
test: overlaping events

bart-krakowski committed:
docs: add the choosing-a-date-time-handling-solution adr

bart-krakowski committed:
feat(react-time/useCalendar): getCurrentTimeMarkerProps

bart-krakowski committed:
feat(react-time/useCalendar): getCurrentTimeMarkerProps

bart-krakowski committed:
feat(react-time/useCalendar): getCurrentTimeMarkerProps

bart-krakowski committed:
refactor: reducer

bart-krakowski committed:
refactor: reducer

bart-krakowski committed:
refactor(useCalendar): types

bart-krakowski committed:
refactor(useCalendar): types

bart-krakowski committed:
refactor: useCalendar reducer

bart-krakowski committed:
refactor: useCalendar reducer

bart-krakowski committed:
refactor: useCalendar reducer

bart-krakowski committed:
feat: add the isToday field

bart-krakowski committed:
feat: add the isToday field

bart-krakowski committed:
feat: add the isToday field

bart-krakowski committed:
docs: useCalendar

bart-krakowski committed:
feat: useCalendar hook

bart-krakowski committed:
cd

bart-krakowski committed:
feat: useCalendar hook

bart-krakowski committed:
feat: useCalendar hook

bart-krakowski committed:
feat: useCalendar hook

bart-krakowski committed:
feat: useCalendar hook

bart-krakowski committed:
test: useCalendar

bart-krakowski committed:
test: useCalendar

bart-krakowski committed:
test: useCalendar

bart-krakowski committed:
feat: custom reducer

bart-krakowski committed:
docs: tsdocs

bart-krakowski committed:
docs: tsdocs

bart-krakowski committed:
docs: tsdocs

bart-krakowski committed:
refactor: useCalendar

bart-krakowski committed:
refactor: useCalendar

bart-krakowski committed:
refactor: useCalendar

bart-krakowski committed:
refactor: change viewMode type

bart-krakowski committed:
refactor: change viewMode type

bart-krakowski committed:
fix: missing days in view

bart-krakowski committed:
refactor: api

bart-krakowski committed:
refactor: api

bart-krakowski committed:
fix: types

bart-krakowski committed:
fix: types

bart-krakowski committed:
fix: types

bart-krakowski committed:
fix: types

bart-krakowski committed:
docs: useCalendar

bart-krakowski committed:
docs: useCalendar

bart-krakowski committed:
fix: types

bart-krakowski committed:
refactor: add the fillMissingDays argument

bart-krakowski committed:
refactor: add the fillMissingDays argument

bart-krakowski committed:
refactor: add the fillMissingDays argument

bart-krakowski committed:
refactor: move utilities to the core package

bart-krakowski committed:
fix: getCurrentTimeMarkerProps

bart-krakowski committed:
docs: update jsdocs

bart-krakowski committed:
test: useCalendar

bart-krakowski committed:
refactor: getCurrentTimeMarkerProps

bart-krakowski committed:
feat: calendar core

bart-krakowski committed:
feat: calendar core

bart-krakowski committed:
feat: calendar core

bart-krakowski committed:
docs: calendar core

bart-krakowski committed:
refactor: useCalendar

bart-krakowski committed:
test: calendar core

bart-krakowski committed:
refactor: useCalendar

bart-krakowski committed:
refactor: useCalendar

bart-krakowski committed:
refactor: useCalendar

bart-krakowski committed:
refactor(useCalendar): types

bart-krakowski committed:
refactor: move calendar-related files to the core directory

bart-krakowski committed:
refactor: update CalendarCoreOptions interface to use default generic type for TEvent

bart-krakowski committed:
refactor: Update CalendarCoreOptions interface in calendar.ts

bart-krakowski committed:
feat: date picker core

bart-krakowski committed:
test: imports

bart-krakowski committed:
feat: date picker core

bart-krakowski committed:
docs: typo

Faithfinder commented:
I'm just keeping a casual eye on the repo, so disregard me If you want, but this feels unintuitive to me. Datepicker inheriting from calendar feels backwards, as Calendar is obviously a bigger concept than date

bart-krakowski commented:
Thanks for your input, @Faithfinder! 
Let me explain. DatePicker has quite a lot in common with the calendar, e.g. moving to subsequent periods, changing the date, etc.
It may be worth creating a separate abstract class that will provide such basic functionality, while the CalendarCore and DatePickerCore classes will extend it by adding specific methods such as event handling and date selection

bart-krakowski committed:
refactor: DatePickerCore

bart-krakowski committed:
refactor: DatePickerCore

Faithfinder commented:
> Thanks for your input, @Faithfinder! Let me explain. DatePicker has quite a lot in common with the calendar, e.g. moving to subsequent periods, changing the date, etc. It may be worth creating a separate abstract class that will provide such basic functionality, while the CalendarCore and DatePickerCore classes will extend it by adding specific methods such as event handling and date selection

Well, I'm not saying it feels _unrelated_, I'm saying it feels _backwards_ :) Anyway, it's not like I dove deep into code or decisions, I'll leave you alone now :)

bart-krakowski committed:
refactor: weekInfoPolyfill

bart-krakowski committed:
refactor: weekInfoPolyfill

bart-krakowski committed:
refactor: weekInfoPolyfill

bart-krakowski committed:
refactor: weekInfoPolyfill

bart-krakowski committed:
refactor: date defaults

bart-krakowski committed:
refactor: date defaults

bart-krakowski committed:
refactor: date defaults

bart-krakowski committed:
refactor: types

bart-krakowski committed:
docs: jsdocs

bart-krakowski committed:
refactor: types

bart-krakowski committed:
refactor: types

bart-krakowski committed:
test: useCalendar

bart-krakowski committed:
refactor: useIsomorphicLayoutEffect

bart-krakowski committed:
chore: resolve conflicts

bart-krakowski committed:
chore: resolve conflicts

bart-krakowski committed:
docs: jsdocs

bart-krakowski committed:
docs: calendar

bart-krakowski committed:
docs: calendar

bart-krakowski committed:
refactor: zonedDateTime

bart-krakowski committed:
refactor: rename startDate to start and endDate to end

bart-krakowski committed:
refactor: remove the getCurrentTimeMarkerProps method

bart-krakowski committed:
refactor: weekInfoPolyfill

bart-krakowski committed:
refactor: add a resources field to event

bart-krakowski committed:
refactor: add a resources field to event

bart-krakowski committed:
refactor: add a resources field to event

bart-krakowski committed:
refactor: add startOf and endOf helpers

bart-krakowski committed:
refactor: remove all the layout specific code

bart-krakowski committed:
refactor: update getDaysNames method to accept weekday parameter

bart-krakowski committed:
refactor: update useCalendar test to mock current date and time

bart-krakowski committed:
refactor: update useCalendar test to mock current date and time

bart-krakowski committed:
test: update useCalendar tests

bart-krakowski committed:
test: update useCalendar tests

bart-krakowski committed:
test: useCalendar

bart-krakowski committed:
refactor: update CalendarCore tests to mock current date and time

bart-krakowski committed:
refactor: Update CalendarCore to use destructuring assignment for datePart

bart-krakowski committed:
feat: add the activeDate property

bart-krakowski committed:
refactor: use compare method

bart-krakowski committed:
feat: add workWeek unit

bart-krakowski committed:
feat: add workWeek unit

bart-krakowski committed:
test: tests reorganisation

bart-krakowski committed:
chore: add the get-week-info-polyfill

bart-krakowski committed:
feat: getWeekInfo polyfill

bart-krakowski committed:
refactor: getEventProps

bart-krakowski committed:
refactor: getEventProps

bart-krakowski committed:
refactor: Update generateDateRange to accept string inputs

bart-krakowski committed:
chore: upgrade polyfill

bart-krakowski committed:
fix: getEventProps

bart-krakowski committed:
fix: getEventProps

bart-krakowski committed:
refactor: update date inputs to accept string format

bart-krakowski committed:
refactor: convert activeDate, currentPeriod to string in useCalendar

bart-krakowski committed:
refactor: startOf and endOf helpers

bart-krakowski committed:
test: calendar core

bart-krakowski committed:
refactor: update useCalendar to accept string inputs for activeDate and currentPeriod

bart-krakowski committed:
refactor: Update get-week-info-polyfill dependency to version 1.0.6

bart-krakowski committed:
refactor: Update startOf and endOf helpers to accept string inputs

bart-krakowski committed:
docs: update architecture

bart-krakowski committed:
# This is a combination of 6 commits.
# This is the 1st commit message:

docs: readme

# This is the commit message #2:

docs: readme

# This is the commit message #3:

docs: update the repo header

# This is the commit message #4:

docs: update the repo header

# This is the commit message #5:

docs: update the repo header

# This is the commit message #6:

docs: update the repo header

bart-krakowski committed:
docs: readme

bart-krakowski committed:
chore: dependencies update

bart-krakowski committed:
chore: dependencies update

bart-krakowski committed:
Merge branch 'TanStack:v0' into v0

bart-krakowski committed:
chore: resolve conflicts

bart-krakowski committed:
Merge pull request #1 from bart-krakowski/chore/vitest-config

Chore/vitest config

bart-krakowski committed:
docs: architecture

docs: architecture

bart-krakowski committed:
Merge branch 'v0-1-0' into docs/readme

bart-krakowski committed:
docs: readme

docs: readme

bart-krakowski committed:
docs: adr choosing a date and time handling solution

docs: add the choosing-a-date-time-handling-solution adr

bart-krakowski committed:
fix: parse util

bart-krakowski committed:
fix: parse util

bart-krakowski committed:
test: parse

test: refactor tests for the parse util

bart-krakowski committed:
fix: parse util

bart-krakowski committed:
Merge branch 'v0-1-0' of github.com:bart-krakowski/time into v0-1-0

bart-krakowski committed:
chore: resolve conflicts

bart-krakowski committed:
chore: resolve conflicts

bart-krakowski committed:
chore: resolve conflicts

bart-krakowski committed:
test: update temporal mocks

bart-krakowski committed:
chore: update the lock file

bart-krakowski committed:
fix: lint

bart-krakowski committed:
fix: lint

bart-krakowski committed:
fix: lint

bart-krakowski committed:
feat: basic example

bart-krakowski committed:
feat: basic example

bart-krakowski committed:
feat: basic example

bart-krakowski committed:
feat: calendar

bart-krakowski committed:
feat: calendar

bart-krakowski committed:
feat: calendar

bart-krakowski committed:
feat: calendar

bart-krakowski committed:
feat: calendar

bart-krakowski committed:
feat: calendar

bart-krakowski committed:
chore: resolve conflicts

bart-krakowski committed:
feat: date picker

bart-krakowski committed:
feat: date picker

bart-krakowski committed:
feat: date picker

bart-krakowski committed:
feat: date picker

bart-krakowski committed:
feat: date picker
